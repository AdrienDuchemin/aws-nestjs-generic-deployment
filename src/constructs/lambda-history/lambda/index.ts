import { Context, SQSHandler, SQSEvent } from 'aws-lambda'
import { DynamoDB } from 'aws-sdk'

export const handler: SQSHandler = async (event: SQSEvent, context: Context) => {
    console.log({ event, context })

    if(!process.env.TABLE_NAME) {
        throw new Error('no table name given')
    }

    const client = new DynamoDB.DocumentClient()

    for(const record of event.Records){
        const body = record.body ? JSON.parse(record.body) : { }
        await client.put({
            TableName: process.env.TABLE_NAME,
            Item: {
                id: record.messageId,
                text: body.text,
            }
        }).promise()
    }
}