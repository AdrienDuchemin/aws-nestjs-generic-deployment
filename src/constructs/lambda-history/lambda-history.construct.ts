import { join } from 'path'
import { Table, AttributeType } from '@aws-cdk/aws-dynamodb'
import { Code } from '@aws-cdk/aws-lambda'
import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources'
import { Queue } from '@aws-cdk/aws-sqs'
import { Construct } from '@aws-cdk/core'
import { Lambda } from '../lambda.construct'

export interface LambdaHistoryProps {
  queue: Queue,
}

export class LambdaHistory extends Lambda {
  constructor(scope: Construct, id: string, props: LambdaHistoryProps) {
    super(scope, id, {
      code: Code.fromAsset(join(__dirname, './lambda')),
      handler: 'index.handler',
      // environment: {
        // RDS_ARN: from props,
      // }
    })

    const { queue } = props
    this.addEventSource(new SqsEventSource(queue))
    
    const table = new Table(this, `${id}Table`, {
      partitionKey: { name: 'id', type: AttributeType.STRING }
    })
    table.grantReadWriteData(this)

    this.addEnvironment('TABLE_NAME', table.tableName)
  }
}