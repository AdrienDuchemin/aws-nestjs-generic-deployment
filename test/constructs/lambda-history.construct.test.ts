import { expect as expectCDK, SynthUtils, countResources, haveResource } from '@aws-cdk/assert'
import { Queue } from '@aws-cdk/aws-sqs'
import { Stack } from '@aws-cdk/core'
import { LambdaHistory, LambdaHistoryProps } from '../../src/constructs/lambda-history/lambda-history.construct'

describe('LambdaHistory', () => {
  let stack: Stack

  beforeAll(()=> {
    stack = new Stack()
    const props: LambdaHistoryProps = {
      queue : new Queue(stack, 'HistoryQueue')
    }
    new LambdaHistory(stack, 'LambdaHistory', props)
  })

  it('should create one Lambda', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1))
  })

  it('should create one Table', () => {
    expectCDK(stack).to(countResources("AWS::DynamoDB::Table", 1))
    // { name: 'id', type: AttributeType.STRING }
    expectCDK(stack).to(haveResource("AWS::DynamoDB::Table",{
      KeySchema: [{ 
        AttributeName: "id",
        KeyType: "HASH",
      }],
    }))
  })

  it('should not have changed', () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})