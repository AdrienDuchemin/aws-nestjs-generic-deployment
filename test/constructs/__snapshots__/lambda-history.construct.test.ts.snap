// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaHistory should not have changed 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05ArtifactHash6241674D": Object {
      "Description": "Artifact hash for asset \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3Bucket3765D8AA": Object {
      "Description": "S3 bucket for asset \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835": Object {
      "Description": "S3 key for asset version \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HistoryQueue16C3A71A": Object {
      "Type": "AWS::SQS::Queue",
    },
    "LambdaHistorylambda1AC469AB": Object {
      "DependsOn": Array [
        "LambdaHistorylambdaServiceRoleDefaultPolicyE9A62A06",
        "LambdaHistorylambdaServiceRole360155C7",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3Bucket3765D8AA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "LambdaHistorylambdaLambdaHistorytable04796A10",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaHistorylambdaServiceRole360155C7",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaHistorylambdaLambdaHistorytable04796A10": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "LambdaHistorylambdaServiceRole360155C7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaHistorylambdaServiceRoleDefaultPolicyE9A62A06": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "HistoryQueue16C3A71A",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "LambdaHistorylambdaLambdaHistorytable04796A10",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaHistorylambdaServiceRoleDefaultPolicyE9A62A06",
        "Roles": Array [
          Object {
            "Ref": "LambdaHistorylambdaServiceRole360155C7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaHistorylambdaSqsEventSourceHistoryQueue48C4543A": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "HistoryQueue16C3A71A",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "LambdaHistorylambda1AC469AB",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
}
`;
